File: bindings/node/src/deck.ts
import path from 'node:path';
import { Buildable } from './buildable';
import { native, type NativeBuildable } from './internal/native';
import { options, string, strings } from './internal/validation';
import { outcome } from './internal/outcome';
import { nativeError } from './errors';
import type { IoMode, Rect } from './types';

export interface DeckOptions {
  baseDir?: string;
  stableId?: string;
  basicIdentity?: readonly ('front' | 'back')[];
}
export interface DeckNoteOptions {
  stableId?: string;
  tags?: readonly string[];
}
export interface DeckBasicOptions extends DeckNoteOptions {
  identityOverride?: { fields: readonly ('front' | 'back')[]; reasonCode: string };
}
export interface DeckClozeOptions extends DeckNoteOptions {
  extra?: string;
}
export interface DeckImageOcclusionOptions extends DeckNoteOptions {
  mode?: IoMode;
  rects: readonly Rect[];
  header?: string;
  backExtra?: string;
  comments?: string;
}
const references = new WeakSet<DeckMediaRef>();
const token = Symbol('deck-media');
export class DeckMediaRef {
  /** @internal Obtained by registering media on a Deck. */
  constructor(
    key: symbol,
    readonly filename: string,
  ) {
    if (key !== token) throw new TypeError('Obtain media from Deck.media');
    references.add(this);
    Object.freeze(this);
  }
}
function identityFields(value: unknown): void {
  strings(value, 'identity fields');
  if (value.some((field) => !['front', 'back'].includes(field)))
    throw new TypeError('Deck identity fields must be front or back');
}
function noteOptions(config: DeckNoteOptions, extraKeys: string[]): void {
  options(config, ['stableId', 'tags', ...extraKeys], 'deck note');
  if (config.stableId !== undefined) string(config.stableId, 'stableId');
  if (config.tags !== undefined) strings(config.tags, 'tags');
}
export class Deck extends Buildable {
  readonly baseDir: string;
  readonly media: {
    addFile: (filename: string) => Promise<DeckMediaRef>;
    addBytes: (name: string, bytes: Uint8Array) => Promise<DeckMediaRef>;
  };
  #deck: InstanceType<ReturnType<typeof native>['NativeDeck']>;
  constructor(
    readonly name: string,
    config: DeckOptions = {},
  ) {
    super();
    string(name, 'name');
    options(config, ['baseDir', 'stableId', 'basicIdentity'], 'deck');
    if (config.stableId !== undefined) string(config.stableId, 'stableId');
    const baseDir = config.baseDir ?? process.cwd();
    string(baseDir, 'baseDir');
    this.baseDir = path.resolve(baseDir);
    if (config.basicIdentity !== undefined) identityFields(config.basicIdentity);
    try {
      this.#deck = new (native().NativeDeck)(
        name,
        JSON.stringify({ stableId: config.stableId, basicIdentity: config.basicIdentity }),
      );
    } catch (error) {
      if (error instanceof Error && error.message.startsWith('{')) outcome(error.message);
      throw error;
    }
    const reference = (input: string) => {
      const result = outcome(input);
      string(result.filename, 'native media filename');
      return new DeckMediaRef(token, result.filename);
    };
    this.media = Object.freeze({
      addFile: async (filename: string) => {
        string(filename, 'filename');
        try {
          return reference(await this.#deck.addMediaFile(path.resolve(this.baseDir, filename)));
        } catch (error) {
          nativeError(error);
        }
      },
      addBytes: async (name: string, bytes: Uint8Array) => {
        string(name, 'name');
        if (!(bytes instanceof Uint8Array)) throw new TypeError('Expected Buffer or Uint8Array');
        try {
          return reference(await this.#deck.addMediaBytes(name, Buffer.from(bytes)));
        } catch (error) {
          nativeError(error);
        }
      },
    });
    Object.freeze(this);
  }
  basic(front: string, back: string, config: DeckBasicOptions = {}): void {
    string(front, 'front');
    string(back, 'back');
    noteOptions(config, ['identityOverride']);
    if (config.identityOverride !== undefined) {
      options(config.identityOverride, ['fields', 'reasonCode'], 'identityOverride');
      identityFields(config.identityOverride.fields);
      string(config.identityOverride.reasonCode, 'reasonCode');
    }
    try {
      outcome(this.#deck.addBasic(front, back, JSON.stringify(config)));
    } catch (error) {
      nativeError(error);
    }
  }
  cloze(text: string, config: DeckClozeOptions = {}): void {
    string(text, 'text');
    noteOptions(config, ['extra']);
    if (config.extra !== undefined) string(config.extra, 'extra');
    try {
      outcome(this.#deck.addCloze(text, JSON.stringify(config)));
    } catch (error) {
      nativeError(error);
    }
  }
  imageOcclusion(image: DeckMediaRef, config: DeckImageOcclusionOptions): void {
    if (!references.has(image)) throw new TypeError('Expected a DeckMediaRef');
    noteOptions(config, ['mode', 'rects', 'header', 'backExtra', 'comments']);
    if (!Array.isArray(config.rects)) throw new TypeError('rects must be an array');
    for (const rect of config.rects) {
      options(rect, ['x', 'y', 'width', 'height'], 'rect');
      for (const key of ['x', 'y', 'width', 'height'])
        if (!Number.isInteger(rect[key]) || Number(rect[key]) < 0 || Number(rect[key]) > 0xffffffff)
          throw new TypeError(`rect.${key} must be a uint32`);
    }
    if (
      config.mode !== undefined &&
      !['hide-all-guess-one', 'hide-one-guess-one'].includes(config.mode)
    )
      throw new TypeError('Invalid image occlusion mode');
    for (const key of ['header', 'backExtra', 'comments'] as const)
      if (config[key] !== undefined) string(config[key], key);
    try {
      outcome(
        this.#deck.addImageOcclusion(
          image.filename,
          JSON.stringify({ tags: [], mode: 'hide-all-guess-one', ...config }),
        ),
      );
    } catch (error) {
      nativeError(error);
    }
  }
  protected nativeProject(): NativeBuildable {
    return this.#deck;
  }
}

File: bindings/python/src/anki_forge/deck.py
from __future__ import annotations

from dataclasses import asdict, dataclass
from collections.abc import Iterable
import json
from pathlib import Path
from typing import Literal

from . import _native
from ._bridge import invoke
from ._buildable import _Buildable
from .diagnostics import ValidationError
from .options import PathInput, absolute_path
from .report import ValidationReport, _diagnostics

BasicIdentityField = Literal["front", "back"]
IoMode = Literal["hide_all_guess_one", "hide_one_guess_one"]


@dataclass(frozen=True)
class BasicIdentityOverride:
    fields: tuple[BasicIdentityField, ...]
    reason_code: str

    def __init__(self, fields: Iterable[BasicIdentityField], reason_code: str) -> None:
        object.__setattr__(self, "fields", tuple(fields))
        object.__setattr__(self, "reason_code", reason_code)


class DeckMediaRef:
    """A Deck media filename, distinct from Project's MediaRef."""

    def __init__(self, handle: _native.NativeDeckMediaRef) -> None:
        if not isinstance(handle, _native.NativeDeckMediaRef):
            raise TypeError("DeckMediaRef requires a registered Deck media handle")
        self._handle = handle

    @property
    def name(self) -> str:
        return self._handle.filename

    def __eq__(self, other: object) -> bool:
        return isinstance(other, DeckMediaRef) and self.name == other.name

    def __hash__(self) -> int:
        return hash(self.name)


class DeckMediaRegistry:
    def __init__(self, handle: _native.NativeDeck, base_dir: Path) -> None:
        self._handle = handle
        self._base_dir = base_dir

    def add_file(self, path: PathInput) -> DeckMediaRef:
        """Register a file under its basename, following Rust Deck semantics."""
        return DeckMediaRef(invoke(self._handle.add_media_file, absolute_path(path, self._base_dir)))

    def add_bytes(self, name: str, data: bytes | bytearray) -> DeckMediaRef:
        if not isinstance(data, (bytes, bytearray)):
            raise ValidationError("media data must be bytes or bytearray")
        return DeckMediaRef(invoke(self._handle.add_media_bytes, name, bytes(data)))

    def get(self, name: str) -> DeckMediaRef | None:
        handle = invoke(self._handle.get_media, name)
        return DeckMediaRef(handle) if handle is not None else None


class Deck(_Buildable):
    """The Rust stock-note Deck, with inferred identity and image geometry checks.

    Basic/Cloze strings follow Rust Deck's HTML semantics. Use Project for
    typed Content inputs and custom note types.
    """

    _handle: _native.NativeDeck

    def __init__(self, name: str, stable_id: str | None = None, *, basic_identity: Iterable[BasicIdentityField] | None = None, base_dir: PathInput | None = None) -> None:
        self._name = name
        self._stable_id = stable_id
        self._base_dir = Path(base_dir or Path.cwd()).expanduser().absolute()
        self._handle = invoke(_native.NativeDeck, name, json.dumps({"stable_id": stable_id, "basic_identity": list(basic_identity) if basic_identity is not None else None}))
        self._media = DeckMediaRegistry(self._handle, self.base_dir)

    @property
    def name(self) -> str:
        return self._name

    @property
    def stable_id(self) -> str | None:
        return self._stable_id

    @property
    def base_dir(self) -> Path:
        return self._base_dir

    @property
    def media(self) -> DeckMediaRegistry:
        return self._media

    def add_basic(self, front: str, back: str, *, stable_id: str | None = None, tags: Iterable[str] = (), identity_override: BasicIdentityOverride | None = None) -> Deck:
        invoke(self._handle.add_basic, front, back, json.dumps({
            "stable_id": stable_id, "tags": list(tags),
            "identity_override": asdict(identity_override) if identity_override is not None else None,
        }, ensure_ascii=False))
        return self

    def add_cloze(self, text: str, *, stable_id: str | None = None, extra: str = "", tags: Iterable[str] = ()) -> Deck:
        invoke(self._handle.add_cloze, text, json.dumps({"stable_id": stable_id, "extra": extra, "tags": list(tags)}, ensure_ascii=False))
        return self

    def add_image_occlusion(self, image: DeckMediaRef, *, rects: Iterable[tuple[int, int, int, int]], stable_id: str | None = None, mode: IoMode = "hide_all_guess_one", header: str = "", back_extra: str = "", comments: str = "", tags: Iterable[str] = ()) -> Deck:
        rectangles = [tuple(rect) for rect in rects]
        if any(len(rect) != 4 or not all(type(v) is int and 0 <= v <= 2**32 - 1 for v in rect) for rect in rectangles):
            raise ValidationError("image occlusion rect values must be u32 integers")
        invoke(self._handle.add_image_occlusion, image._handle, json.dumps({
            "stable_id": stable_id, "mode": mode, "rects": rectangles, "header": header,
            "back_extra": back_extra, "comments": comments, "tags": list(tags),
        }, ensure_ascii=False))
        return self

    def validate(self) -> ValidationReport:
        return ValidationReport(_diagnostics(json.loads(invoke(self._handle.validate))))
